### DISCUSSION TITLE
High polling rate (>1000hz) input devides overwhealm the input handler/game engine

### DISCUSSION THREAD

**Discussion by lisekilis** [ppy/osu#27718] [Q&A]
### Type

Crash to desktop

#### @ 1000hz
    Raw input Off: Normal behavior, basically no input delay
    Raw input On: Same story, but the FPS is slightly lower
#### @ 2000hz
    Raw input Off: Basically normal behavior, but the input process time is higher. When the mouse moves really fast, some inputs get ignored.
    Raw input On: Same thing, but the process time is higher, and the excessive inputs get delayed instead of ignored
#### @ 8000hz
    Raw input Off: The game again ignores inputs, but this time the cursor jumps a lot more, and it's very noticeable.
    Raw input On: The faster the mouse moves, the more the game struggles. At higher speeds, it just suppresses all inputs.

### Here's my systems information regardless
```
Mouse used: Corsair M65 Ultra (wired)
                                ..,   lisek@DESKTOP-FOXv2
                    ....,,:;+ccllll   -------------------
      ...,,+:;  cllllllllllllllllll   OS: Windows 10 Pro x86_64
,cclllllllllll  lllllllllllllllllll   Host: Micro-Star International Co., Ltd. MS-7C75
llllllllllllll  lllllllllllllllllll   Kernel: 10.0.19045
llllllllllllll  lllllllllllllllllll   Uptime: 3 hours, 27 mins
llllllllllllll  lllllllllllllllllll   Packages: 2 (scoop)
llllllllllllll  lllllllllllllllllll   Shell: bash 5.2.15
llllllllllllll  lllllllllllllllllll   Resolution: 1920x1080
                                      DE: Aero
llllllllllllll  lllllllllllllllllll   WM: Explorer
llllllllllllll  lllllllllllllllllll   WM Theme: Custom
llllllllllllll  lllllllllllllllllll   CPU: Intel i5-10600KF (12) @ 4.110GHz
llllllllllllll  lllllllllllllllllll   GPU: Caption
llllllllllllll  lllllllllllllllllll   GPU: NVIDIA GeForce RTX 3060 Ti
`'ccllllllllll  lllllllllllllllllll   GPU Memory: 16771MiB / 32689MiB
       `' \*::  :ccllllllllllllllll   
                       ````''*::cll
                                 ``
```
####
I can provide more information if needed

### Screenshots or videos

https://github.com/ppy/osu/assets/101191226/2d08582f-7ec6-4eef-a692-e372a41f7c02

https://github.com/ppy/osu/assets/101191226/18b2627e-82f1-4b94-aca5-a8c10614e5c1

https://github.com/ppy/osu/assets/101191226/22dd49f9-1258-41e5-b5c2-d1cee6d8ae44

https://github.com/ppy/osu/assets/101191226/128f24ed-bdac-4689-9822-62f5b67fdf26

The 8k hz polling rate videos are really low quality because of the 10mb size restriction

[whole testing video @ higher quality](https://drive.google.com/file/d/1p1iyojH4Tk5ZTk7bWRa8B7-D-Ic5dG2k/view?usp=sharing)

### Version

2024.312.1-lazer

### Logs

[compressed-logs.zip](https://github.com/ppy/osu/files/14736401/compressed-logs.zip)

**Comment by peppy** [comment]
See that white area? That's other apps on your system struggling, not osu!. They are blocking the windows main loop which is why you're seeing the issue manifest in osu!, but it willhappen across many other apps.

Solution: reduce polling rate, 8000hz is not going to benefit you.

**Reply by lisekilis** [reply]
I wonder what application might be doing that since I have tried turning off everything that is not needed and the issue persisted
I'll keep testing

**Reply by lisekilis** [reply]
After further testing, I discovered some programs that cause issues when using an 8khz mouse (especially with raw input enabled):

- Corsair iCUE (app only, drivers work fine)
- Discord, Steam, Spotify (potential Electron framework issue?)

This issue seems to be caused by a combination of factors, as some games work even with iCUE and Discord running in the background.

Examples of working games: CS:GO, Valorant (according to Valorant patch notes 3.07: [playvalorant.com](https://playvalorant.com/en-gb/news/game-updates/valorant-patch-notes-3-07/)), and apparently Overwatch (though I couldn't test it since the game has reached EOL).

https://github.com/ppy/osu/assets/101191226/7d6a1e8c-c4cd-41b9-835b-b815961e30f5

(tested with both discord and icue in the background)

**Reply by morethanhater** [reply]
> After further testing, I discovered some programs that cause issues when using an 8khz mouse (especially with raw input enabled):
> 
> * Corsair iCUE (app only, drivers work fine)
> * Discord, Steam, Spotify (potential Electron framework issue?)
> 
> This issue seems to be caused by a combination of factors, as some games work even with iCUE and Discord running in the background.

Just bought a new mouse and faced the stuttering problem at anything higher than 1k polling rate.
Experimented a bit and it's true that some apps affects performance in Osu and, perhaps, at other apps as well.
It's basically impossible to play Osu with Steam or/and Discord in the background.
Although, Stalker 2 felt fine even at 8k with enabled Discord and Steam overlay.

In osu, if I disable both steam and discord overlay:
2k polling rate gives me 0.8ms delay (bottom right corner stat with unlimited FPS option selected at game's settings) and so far no stutter for like half an hour gameplay at 6-7 stars maps. 
At 4k polling rate it gives me 1.1ms delay, but sometimes it drops a huge stutter, it even may be quite calm part of the map where there's no extensive pointer movements.
At 8k polling rate it's  unplayable, delay is more than 2ms and stutters a super lot.

For reference, my CPU is Ryzen 5 9600x and GPU is nvidia 4070ti.
And yes, Raw input is enabled.

In the end of the day, if other games works fine at even 8k polling rate + enabled Steam and Discord, but Osu doesn't, the only reasonable explaination would be that Osu needs optimization on this part.

**Comment by adim2222** [comment]
can confirm this issue on windows 10 with 9070xt + 9800x3d
this is definitely not other apps fault

**Reply by peppy** [reply]
please report back on a fresh windows 11 install, with nothing else installed.

**Comment by Rothes** [comment]
I have 9070xt + 9700X with Windows 11, a 8K mouse. Some background apps will make the game stutter hugely, like `Chrome`, `Steam`, `Razer Synapse`, `Wallpaper Engine` or any app that may use GPU. Without those app running, there's almost no lag, Input runs on 600+FPS. `vscode` won't introduce stutters so I think it's not Electron related.

I have found out AMD display driver `amdkmdag.sys` + `dxgkrnl.sys` took too much execution time.
It looks like AMD should fix it. You can check it with LatencyMon software, and report it to AMD if you have same experience with me. 

The other games work smoothly because of they buffer the inputs.

**Comment by morethanhater** [comment]
Refer to this post: https://github.com/ppy/osu/discussions/27718#discussioncomment-14397403
Bumping it like this as comments to older posts are often being missed

**Reply by Rothes** [reply]
> Refer to this post: [#27718 (reply in thread)](https://github.com/ppy/osu/discussions/27718#discussioncomment-14397403) Bumping it like this as comments to older posts are often being missed

I believe the other doesn't lag is due to the introduce of input buffer. For me, with it enabled, the game doesn't lag, but I may get a even higher input latency than 1/2khz.

For both the latency and performance, I think the only thing we can do is to optimise our own system, especially background softwares.
Check https://github.com/Aetopia/NoSteamWebHelper, you can disable steam browser, so there's almost no impact from Steam.

**Reply by morethanhater** [reply]
> For me, with it enabled

What software do you use for this? Seems like not a native Osu setting.
Thanks. Although Steam Web Helper impact seems to be almost not noticable. What is noticable, is, Discord overlay.
By disabling it, input delay shortens up to 3 times. Discord here is the major culprit. 

I'd like to emphasise once again that discord overlay doesn't increase input lag when mouse is in 1k polling rate state. Increased input delay as well as heavy stutters only occurs when you use mouse's polling rate higher than 1k.
<img width="1260" height="641" alt="image" src="https://github.com/user-attachments/assets/7025d5b6-dd7a-4725-85c1-0adcd094f7d6" />

**Reply by lisekilis** [reply]
When I get home I shall test it with the new discord overlay since it doesn't hook into the game, but just add an overlay over everything when the game is in view and not in exclusive fullscreen (which is still suboptimal as osu! likes exclusive fullscreen the best)

**Reply by lisekilis** [reply]
Alright. after further testing I have confirmed that using the new discord overlay seems to have a smaller impact, however

Having anything chrome related open in the background (especially Spotify and Steam and Chrome itself) has the biggest impact

Discord by itself also affects it but way less that something like the programs mentioned above

**Reply by lisekilis** [reply]
I have also upgraded my hardware since last time and even with an r7 9800x3d this issue overwhealms the input handler down to about 600 hz depending on what's going on with raw input on and stuff open in the background
way better than the 4fps i was able to achieve previously but still suboptmal.

**Comment by adim2222** [comment]
did the same thing on linux, for some reason works perfectly fine here.
